### ISSUE TITLE
Provide a flow which finishes the stream after not seeing an event for some time.

### ISSUE THREAD

**Issue by markusthoemmes** [akka/akka#24951]
There is a multitude of time-aware stages in akka-streams. For this specific use-case there are:

* `takeWhile(time)`: Close the stream after consuming elements for `time`.
* `idleTimeout(time)`: **Fails** the stream if there is more than `time` between two elements.

My specific use-case is as follows: I have a stream of roughly in-order data. Multiple producers produce to a topic. The data needs to be grouped by their timestamp (based on a given time-frame granularity).

Since I cannot know when the last element of a specific timeframe will arrive (one producer might emit less elements than others and thus could be stuck behind many messages from another producer) I'd like to keep my stream open until no new element arrives for `time`. It should close as usual and emit the aggregate I computed.

It's basically the same flow as `idleTimeout` but without failing the stream. Do you think something like this is worthwhile having in general? If yes: Shall we  add a new or add a parameter to all the `Timeout` flows stating if they should error or not?

**Comment by ktoso** [comment]
I propose to add instead the missing piece that is onError => onComplete, the stage should be something like mapError however for either termination signal. 

We have

`  def mapError(pf: PartialFunction[Throwable, Throwable]): Repr[Out]`
`  def recoverWithRetries[T >: Out](attempts: Int, pf: PartialFunction[Throwable, Graph[SourceShape[T], NotUsed]]): Repr[T] =`

but there is no way to make such error a completion. What we'd need is a `def recoverTermination(PartialFunction[Try[Done], Try[Done]])`.

Done is Complete
some exception is onError

this can handle all cases; error => complete, complete => error, and the identity function as well.


-1 for adding parameters to idleTimeout and friends.
+1 for adding the new operator.
+0.5 for also adding the simplified one that makes `recoverErrorToComplete` since I've seen it requested a number of times

**Comment by markusthoemmes** [comment]
@ktoso Fine for me as well (although I feel like both is worth having).

A few questions:

### First:
Would `recoverWithRetries(0, case _ => Source.empty)` work in theory?

### Second:
> Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements. This stage can recover the failure signal, but not the skipped elements, which will be dropped.

That sounds a bit scary. Are deeper cuts needed into how errors are handled by akka-streams to do the fix you suggested?

**Comment by ktoso** [comment]
> >    Since the underlying failure signal onError arrives out-of-band, it might jump over existing elements. This stage can recover the failure signal, but not the skipped elements, which will be dropped.
> 
> That sounds a bit scary. Are deeper cuts needed into how errors are handled by akka-streams to do the fix you suggested?

All reactive streams implementations work like this; This is not to be changed.

If you care about keeping *errors* in stream order, then they must be elements.
Stream failure is like an "OOM", it's not something you should continiously do and recover from, it's really for "whooop! bad stuff happened, we're in a possibly bad state"

**Comment by ktoso** [comment]
> Would recoverWithRetries(0, case _ => Source.empty) work in theory?

I think so... try it. (can't remember if 1 or 0 there, see scaladoc)

**Comment by markusthoemmes** [comment]
> If you care about keeping errors in stream order, then they must be elements.
Stream failure is like an "OOM", it's not something you should continiously do and recover from, it's really for "whooop! bad stuff happened, we're in a possibly bad state"

That's why I'd rather add new time-based operators which allow for this use-case instead of throwing an error, catch it and cover it up. Does "+1 for adding the new operator." mean you'd be in favor of that as well or was that a "+1 towards adding `recoverTermination`?

**Comment by ktoso** [comment]
> That's why I'd rather add new time-based operators which allow for this use-case instead of throwing an error, catch it and cover it up. 

I see what you mean; Bear in mind that the complete on timeout is racy anyway, so it's not like you gained full determinism. You are right though that it may cause dropping of an element... Though if we ensured the two are always fused.

I would like the `recoverTermination` in any case I guess.
Though your argument about timing may indeed be helpful... How about adding a parameter that is `exception => Try(ex)` which could wrap in a Failed if we want failure, and `Try(Done)` if we want completion after all

**Comment by ktoso** [comment]
More opinions welcome here @patriknw @johanandren

**Comment by johanandren** [comment]
I think a complete-after-idle would be nice. 

Not sure I think the transform-failure-to-completion one is a good idea since I think that having that may make people think ordering is guaranteed.

**Comment by ktoso** [comment]
Ok, so the overloads. With which API. I proposed one above with Try, or a boolean flag, WDYT

**Comment by patriknw** [comment]
BTW, I don't understand why we deprecated `recoverWith` in favor of `recoverWithRetries`. I'd say that in many cases you don't care about the retry counter.

**Comment by ktoso** [comment]
Yeah that was my proposal initially. Esp the fusing aspect I thought makes it safe enough dropping wise, since they are racy anyway. 

Can we then start with the new operator and see how it does? (the one recovering into a completion)

**Comment by markusthoemmes** [comment]
> I'm not sure it's great to add overloads with boolean parameter to all of those compared to composing with .recoverWith or perhaps add a new recoverWithCompleted.
You typically place that immediately after the other stage that emits the failure and then they are fused and there is no real surprise with regards to dropped elements.

It sounds a bit dangerous to me to rely on fusion here. That breaks the `Materializer` abstraction a "bit", does it?

**Comment by ktoso** [comment]
As I mentioned, it is timing stages — they are racy by definition anyway to begin with.

**Comment by markusthoemmes** [comment]
Ha, that's a fair point! 🎉 

One last question regarding "catching the error signal" (I'm not super deep into reactive streams, please bear with me): Will the error signal be signaled downstream strictly?

As in: Is the usual flow:
1. Flow fails, propagates failure downstream
2. Failure reaches sink, fails the stream
3. Failure signal walks upstream

And now it'd be:
1. Flow fails, propagates failure downstream
2. Failure reaches recover flow, is recovered to completion
3. Completion signal reaches sink, completes stream
4. Completion signal walks upstream

**Comment by johanandren** [comment]
I think you could rather describe it as: 

1. stage fails
2.  in some order
    * upstream is cancelled 
    * failure is propagated downstream.

**Comment by hepin1989** [comment]
@markusthoemmes Do you think the `statefulMap` can help?

**Comment by hepin1989** [comment]
@patriknw https://github.com/akka/akka/pull/20137 deprecated the `recoverWith`.

**Comment by patriknw** [comment]
@He-Pin Same question here. You added the exception subclasses in https://github.com/akka/akka/pull/31640, but does that fix this issue, or is our conclusion that this feature request is not needed and we can close this ticket?

**Comment by He-Pin** [comment]
@johanandren I think this issue can be closed

**Comment by johanandren** [comment]
Dedicated exception for idle timeout + completeOnError should cover it. 👍
